pilot/pkg/model/validation.go

```diff
@@ -1476,19 +1476,47 @@ func ValidateServiceRole(name, namespace string, msg proto.Message) error {
 			if len(constraint.Values) == 0 {
 				errs = appendErrors(errs, fmt.Errorf("at least 1 value must be specified for constraint %d in rule %d", j, i))
 			}
+			if hasExistingFirstClassFieldInRole(constraint.Key, rule) {
+				errs = appendErrors(errs, fmt.Errorf("cannot define %s for rule %d because a similar first-class field has been defined", constraint.Key, i))
+			}
 		}
 	}
 	return errs
 }
 
+// Returns true if the user defines a constraint that already exists in the first-class fields, false
+// if none has overlapped.
+func hasExistingFirstClassFieldInRole(constraintKey string, rule *rbac.AccessRule) bool {
+	// Same as authz.attrDestPort
+	// Cannot use authz.attrDestPort since there is a import cycle. However, these constants can be
+	// defined at another place and both authz and model package can access them.
+	const attrDestPort = "destination.port"
+	if constraintKey == attrDestPort && len(rule.Ports) > 0 {
+		return true
+	}
+	if constraintKey == attrDestPort && len(rule.NotPorts) > 0 {
+		return true
+	}
+	return false
+}
+
+// ValidateServiceRoleBinding checks that ServiceRoleBinding is well-formed.
 func checkServiceRoleBinding(in *rbac.ServiceRoleBinding) error {
 	var errs error
 	if len(in.Subjects) == 0 {
 		errs = appendErrors(errs, fmt.Errorf("at least 1 subject must be specified"))
 	}
 	for i, subject := range in.Subjects {
-		if len(subject.User) == 0 && len(subject.Group) == 0 && len(subject.Properties) == 0 {
-			errs = appendErrors(errs, fmt.Errorf("at least 1 of user, group or properties must be specified for subject %d", i))
+		if isFirstClassFieldEmpty(subject) {
+			errs = appendErrors(errs, fmt.Errorf("at least 1 first class field (e.g. properties) must be specified for subject %d", i))
```

Comment:
`first class field` is not very understandable for customer, maybe just say we don't allow empty Subject?

Reply:
Edited. Thanks!